### PR TITLE
[hailctl] update to dataproc 2.1.33

### PR DESCRIPTION
**Pull request by danking**
CHANGELOG: `hailctl dataproc` now creates clusters using Dataproc version 2.1.33. It previously used version 2.1.2.

Dataproc verifications:
1. Memory available after startup: `mem avail: 42959 of 52223 MiB`. OK.
2. SparkMonitor still shows up.
3. Native code (`hl.identity_by_descent`) still works.

The Dataproc version list only keeps the most recent five, sadly. Lucky for us, archive.org happens to have a capture with 2.1.2. I also requested a capture of the current state of the page.

- https://cloud.google.com/dataproc/docs/concepts/versioning/dataproc-release-2.1

- 2023-03-07: https://web.archive.org/web/20230307225815/https://cloud.google.com/dataproc/docs/concepts/versioning/dataproc-release-2.1

- 2023-12-12: https://web.archive.org/web/20231212001716/https://cloud.google.com/dataproc/docs/concepts/versioning/dataproc-release-2.1

| Component | 2.1.33-debian11 | 2.1.2-debian11 |
| --- | --- | --- |
| Apache Atlas | 2.2.0 | 2.2.0 |
| Apache Flink | 1.15.4 | 1.15.0 |
| Apache Hadoop | 3.3.6 | 3.3.3 |
| Apache Hive | 3.1.3 | 3.1.3 |
| Apache Hive WebHCat | 3.1.3 | 3.1.3 |
| Apache Hudi | 0.12.3 | 0.12.0 |
| Apache Kafka | 3.1.0 | 3.1.0 |
| Apache Pig | 0.18.0-SNAPSHOT | 0.18.0-SNAPSHOT | | Apache Spark | 3.3.2 | 3.3.0 |
| Apache Sqoop v1 | 1.5.0-SNAPSHOT | 1.5.0-SNAPSHOT | | Apache Sqoop v2 | N/A | 1.99.6 |
| Apache Tez | N/A | 0.10.1 |
| Cloud Storage Connector | hadoop3-2.2.18 | hadoop3-2.2.9 | | Conscrypt | 2.5.2 | 2.5.2 |
| Docker | 20.10 | 20.10 |
| Hue | 4.10.0 | 4.10.0 |
| Java | 11 | 11 |
| JupyterLab Notebook | 3.4 | 3.4 |
| Oozie | 5.2.1 | 5.2.1 |
| Trino | 376 | 376 |
| Python | Python 3.10 | Python 3.10 |
| R | R 4.1 | R 4.1 |
| Ranger | 2.2.0 | 2.2.0 |
| Scala | 2.12.18 | 2.12.14 |
| Solr | 9.0.0 | 9.0.0 |
| Zeppelin Notebook | 0.10.1 | 0.10.1 |
| Zookeeper | 3.8.3 | 3.8.0 |

Here's a diff. I think the things that may affect us are:

1. Scala patch version
4. Cloud Storage Connector patch version (this is the motivation for this change)
5. Spark patch version
6. Total daemon memory use at start-up.

```diff
diff -u /tmp/b /tmp/a
--- /tmp/b	2023-12-11 19:20:34
+++ /tmp/a	2023-12-11 19:22:42
@@ -1,17 +1,17 @@
-Component	2.1.2-debian11
+Component	2.1.33-debian11
 Apache Atlas	2.2.0
-Apache Flink	1.15.0
-Apache Hadoop	3.3.3
+Apache Flink	1.15.4
+Apache Hadoop	3.3.6
 Apache Hive	3.1.3
 Apache Hive WebHCat	3.1.3
-Apache Hudi	0.12.0
+Apache Hudi	0.12.3
 Apache Kafka	3.1.0
 Apache Pig	0.18.0-SNAPSHOT
-Apache Spark	3.3.0
+Apache Spark	3.3.2
 Apache Sqoop v1	1.5.0-SNAPSHOT
-Apache Sqoop v2	1.99.6
-Apache Tez	0.10.1
-Cloud Storage Connector	hadoop3-2.2.9
+Apache Sqoop v2	N/A
+Apache Tez	N/A
+Cloud Storage Connector	hadoop3-2.2.18
 Conscrypt	2.5.2
 Docker	20.10
 Hue	4.10.0
@@ -22,7 +22,7 @@
 Python	Python 3.10
 R	R 4.1
 Ranger	2.2.0
-Scala	2.12.14
+Scala	2.12.18
 Solr	9.0.0
 Zeppelin Notebook	0.10.1
-Zookeeper	3.8.0
\ No newline at end of file
+Zookeeper	3.8.3
```

### CODE REVIEW

**Comment by danking** [comment]
On an n1-highmem-8, this is the RAM in use after startup. About 200MiB lower overhead than [the last time I did this](https://github.com/hail-is/hail/issues/13960#issuecomment-1836844790).
```
Dec 12 15:20:34 dk-m post-hdfs-startup-script[10260]: + echo 'All done'
Dec 12 15:20:34 dk-m post-hdfs-startup-script[10260]: All done
Dec 12 15:20:42 dk-m earlyoom[6529]: mem avail: 42959 of 52223 MiB (82.26%), swap free:    0 of    0 MiB ( 0.00%)
```

**Comment by danking** [comment]
SparkMonitor still works.

**Comment by danking** [comment]
Native code still works (tested by running `hl.identity_by_descent`).

**Comment by danking** [comment]
Lints now fixed. I expect it to pass.

File: hail/build.gradle [hunk 1]
```diff
@@ -42,12 +42,12 @@ tasks.withType(JavaCompile) {
 project.ext {
     breezeVersion = "1.1"
 
-    sparkVersion = System.getProperty("spark.version", "3.3.0")
+    sparkVersion = System.getProperty("spark.version", "3.3.2")
 
-    if (sparkVersion != "3.3.0") {
-        project.logger.lifecycle("WARNING: Hail primarily tested with Spark 3.3.0, use other versions at your own risk.")
+    if (sparkVersion != "3.3.2") {
+        project.logger.lifecycle("WARNING: Hail primarily tested with Spark 3.3.2, use other versions at your own risk.")
```
Comment (daniel-goldstein) [inline hail/build.gradle:48]: This isn't exactly accurate though, the pinned requirements say we're using `pyspark==3.3.3`. I think we should either omit the patch version or use what is in our `pinned-requirements.txt`.

**Review by daniel-goldstein — DISMISSED** [review]

File: hail/build.gradle (hunk 1 above)
Comment (danking) [reply hail/build.gradle:48]: Good catch. Hmm. I really want us to test with the version of Spark that is used in Dataproc but I don't want to put a pin in the public requirements. Is it possible to constrain the pinned pyspark without the addition of a new requirements file or should I add a `hail/python/constraints-requirements.txt` or w/e?

Comment (daniel-goldstein) [reply hail/build.gradle:48]: Ya I think if you wanted to do that and keep `requirements.txt` open then you would have to add another requirements file with `pyspark==3.3.2`. Or, if what you're really after is just keeping the wheel open then you could tweak the code already in `setup.py` to unconditionally open up the `pyspark` requirement and keep `requirements.txt` pinned.

**Comment by danking** [comment]
@daniel-goldstein OK, so:

1. There is a bug in ruff, so I pinned to a lower version. Ruff issue: https://github.com/astral-sh/ruff/issues/9133
2. I pinned to pyspark 3.3.2 using a new constraint file. See this commit: https://github.com/hail-is/hail/pull/14093/commits/a1d661c62c928fdb60b906cd088958a46129ffc9
3. Regenerating the requirements files pulled in a new upright which found several issues. I fixed them all in a least effort way. https://github.com/hail-is/hail/pull/14093/commits/2cb2c947d6f140b75315549cf42395b6bf3cc3c2


Take a look at (3) in particular. If anything seems funky/annoying, LMK and I can explain and/or we can change.

File: hail/python/hailtop/aiocloud/aioazure/fs.py [hunk 2]
```diff
@@ -400,8 +400,7 @@ async def generate_sas_token(
             expiry=datetime.utcnow() + valid_interval)
         return token
 
-    @staticmethod
-    def parse_url(url: str) -> AzureAsyncFSURL:
+    def parse_url(self, url: str) -> AzureAsyncFSURL:
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiocloud/aioazure/fs.py:403]: Why can't this be static?

File: hail/python/hailtop/aiocloud/aiogoogle/client/base_client.py [hunk 3]
```diff
@@ -7,8 +7,6 @@
 
 
 class GoogleBaseClient(CloudBaseClient):
-    _session: GoogleSession
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiocloud/aiogoogle/client/base_client.py:10]: Love this

File: hail/python/hailtop/aiocloud/aiogoogle/client/storage_client.py [hunk 4]
```diff
@@ -132,7 +132,7 @@ async def __aexit__(self,
 
 
 class ResumableInsertObjectStream(WritableStream):
-    def __init__(self, session: GoogleSession, session_url: str, chunk_size: int):
+    def __init__(self, session: BaseSession, session_url: str, chunk_size: int):
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiocloud/aiogoogle/client/storage_client.py:135]: What's the reason for this change? This is the Google storage client, I feel like it needs a Google session

File: hail/python/hailtop/aiotools/fs/fs.py [hunk 5]
```diff
@@ -133,10 +133,7 @@ class AsyncFS(abc.ABC):
     FILE = 'file'
     DIR = 'dir'
 
-    @property
-    @abc.abstractmethod
-    def schemes(self) -> Set[str]:
-        pass
+    schemes: ClassVar[Set[str]]
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiotools/fs/fs.py:136]: Curious, will pyright complain if a subclass doesn't set this?

File: hail/python/hailtop/fs/router_fs.py [hunk 6]
```diff
@@ -117,7 +117,7 @@ def isatty(self):
     def readable(self):
         return False
 
-    def readline(self, size=-1):
+    def readline(self, size: Optional[int] = -1):
```
Comment (daniel-goldstein) [inline hail/python/hailtop/fs/router_fs.py:120]: What does `size=None` mean?

File: hail/python/test/hail/fs/test_worker_driver_fs.py [hunk 7]
```diff
@@ -135,9 +136,14 @@ def test_can_access_public_blobs():
 
 @run_if_azure
 @fails_local_backend
-def test_qob_can_use_sas_tokens():
+@pytest.mark.asyncio
+async def test_qob_can_use_sas_tokens():
```
Comment (daniel-goldstein) [inline hail/python/test/hail/fs/test_worker_driver_fs.py:140]: pytest should be able to pick up on async functions now without the annotation. Is this not true?

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
Changes look great, just a few questions

**Review by danking — COMMENTED** [review]
Alright, this became a lot bigger but I think it's better.

I killed *Session in favor of BaseSession and Session. The only cloud-specific functionality is the credentials, a session is just a thing that makes HTTP requests.

AsyncFS now has static method parse_url and schemes. For this to happen, RouterAsyncFS became fully lazy (what did passing it filesystems mean anyway?). You can't pass it filesystems anymore.

File: hail/python/test/hail/fs/test_worker_driver_fs.py (hunk 7 above)
Comment (danking) [reply hail/python/test/hail/fs/test_worker_driver_fs.py:140]: I didn't think that was true, but apparently you can turn pytest-asyncio into auto mode to let you do that. I don't know how to do that and it doesn't seem to be documented well: https://pytest-asyncio.readthedocs.io/en/latest/concepts.html#auto-mode

File: hail/python/hailtop/aiocloud/aiogoogle/client/storage_client.py (hunk 4 above)
Comment (danking) [reply hail/python/hailtop/aiocloud/aiogoogle/client/storage_client.py:136]: I didn't kill GoogleStorageClient b/c I had to add like five lines to each client.

File: hail/python/hailtop/aiocloud/aioazure/fs.py [hunk 8]
```diff
@@ -368,6 +368,10 @@ def __init__(self, *, credential_file: Optional[str] = None, credentials: Option
         self._credential = credentials.credential
         self._blob_service_clients: Dict[Tuple[str, str, Union[AzureCredentials, str, None]], BlobServiceClient] = {}
 
+    @staticmethod
+    def schemes() -> Set[str]:
+        return {'hail-az', 'https'}
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiocloud/aioazure/fs.py:373]: Side-note: would you mind making an issue to drop support for `hail-az`? I believe we just need to update our terraform to use `https` in the global config and then remove it from the code base.

File: hail/python/hailtop/aiocloud/aiogoogle/client/base_client.py [hunk 9]
```diff
@@ -1,16 +1,30 @@
...
         if session is None:
-            session = GoogleSession(**kwargs)
-        super().__init__(base_url, session, rate_limit=rate_limit)
+            session = Session(
+                credentials=GoogleCredentials.from_args(credentials, credentials_file),
+                params=params,
+                **kwargs
+            )
+        elif credentials_file is not None or credentials is not None:
+            raise ValueError('Do not provide credentials_file or credentials when session is None')
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiocloud/aiogoogle/client/base_client.py:28]: Same here: *not* None

File: hail/python/hailtop/aiocloud/aiogoogle/client/bigquery_client.py [hunk 10]
```diff
@@ -108,9 +111,21 @@ async def __anext__(self):
                 raise StopAsyncIteration
 
 
-class GoogleBigQueryClient(GoogleBaseClient):
-    def __init__(self, project, **kwargs):
-        super().__init__(f'https://bigquery.googleapis.com/bigquery/v2/projects/{project}', **kwargs)
+class GoogleBigQueryClient(CloudBaseClient):
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiocloud/aiogoogle/client/bigquery_client.py:114]: Why not subclass `GoogleBaseClient`?

File: hail/python/hailtop/aiocloud/aiogoogle/credentials.py [hunk 11]
```diff
@@ -52,6 +52,19 @@ def __init__(self,
...
+        credentials: Optional[Union['GoogleCredentials', AnonymousCloudCredentials]] = None,
+        credentials_file: Optional[str] = None,
+    ) -> Union['GoogleCredentials', AnonymousCloudCredentials]:
+        if credentials is None:
+            if credentials_file:
+                return GoogleCredentials.from_file(credentials_file)
+            return GoogleCredentials.default_credentials()
+        if credentials_file is not None:
+            raise ValueError('Do not provide credentials_file and credentials.')
+        return credentials
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiocloud/aiogoogle/credentials.py:66]: Same comment here about passing back the credentials arg

File: hail/python/hailtop/aiotools/fs/fs.py [hunk 12]
```diff
@@ -133,18 +133,19 @@ class AsyncFS(abc.ABC):
     FILE = 'file'
     DIR = 'dir'
 
-    @property
+    @staticmethod
     @abc.abstractmethod
-    def schemes(self) -> Set[str]:
+    def schemes() -> Set[str]:
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiotools/fs/fs.py:138]: Great change

File: hail/python/hailtop/aiotools/router_fs.py [hunk 13]
```diff
@@ -135,9 +152,11 @@ async def rmtree(self,
...
-        for fs in self._filesystems:
-            await fs.close()
+        await self._exit_stack.aclose()
 
     def copy_part_size(self, url: str) -> int:
         fs = self._get_fs(url)
         return fs.copy_part_size(url)
+
+
+print(RouterAsyncFS().schemes)
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiotools/router_fs.py:162]: Remove?

File: hail/python/test/hail/fs/test_worker_driver_fs.py (hunk 7 above)
Comment (daniel-goldstein) [reply hail/python/test/hail/fs/test_worker_driver_fs.py:140]: I believe I've done so in the batch `pytest.ini` file

File: hail/python/test/hailtop/inter_cloud/test_fs.py [hunk 14]
```diff
@@ -23,12 +23,10 @@ async def filesystem(request) -> AsyncIterator[Tuple[asyncio.Semaphore, AsyncFS,
...
-            fs = RouterAsyncFS(filesystems=[
-                LocalAsyncFS(thread_pool),
-                GoogleStorageAsyncFS(),
-                S3AsyncFS(thread_pool),
-                AzureAsyncFS()
-            ])
+            fs = RouterAsyncFS(
+                local_kwargs={'thread_pool': thread_pool},
+                s3_kwargs={'thread_pool': thread_pool},
+            )
```
Comment (daniel-goldstein) [inline hail/python/test/hailtop/inter_cloud/test_fs.py:29]: This does greatly simplify the ownership story.

**Review by daniel-goldstein — DISMISSED** [review]

File: hail/python/hailtop/fs/router_fs.py (hunk 6 above)
Comment (danking) [reply hail/python/hailtop/fs/router_fs.py:120]: Python folks got back to me. This is actually intentional. Python supports None as a synonym of -1. It does not support this behavior on the TextIOBase subclass, so Python is internally inconsistent. 🤷

File: hail/python/test/hail/fs/test_worker_driver_fs.py (hunk 7 above)
Comment (danking) [reply hail/python/test/hail/fs/test_worker_driver_fs.py:140]: Yeah, you're definitely right. I have deleted every occurrence of pytest.mark.asyncio from the codebase and inserted pytest.ini in `ci/test` and `monitoring/test`.

File: hail/python/hailtop/aiocloud/aioazure/client/pricing_client.py [hunk 15]
```diff
@@ -1,16 +1,11 @@
...
-
 from .base_client import AzureBaseClient
 
 
 class AzurePricingClient(AzureBaseClient):
-    def __init__(self):
-        session = AzureSession(credentials=AnonymousCloudCredentials())
-        super().__init__('https://prices.azure.com/api/retail', session=session)
+    def __init__(self, **kwargs):
+        super().__init__('https://prices.azure.com/api/retail', **kwargs)
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/python/hailtop/aiocloud/aioazure/client/pricing_client.py:8]: Correct. Reverted.

File: hail/python/hailtop/aiocloud/aioazure/credentials.py [hunk 16]
```diff
@@ -59,6 +59,20 @@ async def close(self) -> None:
...
+        credentials_file: Optional[str] = None,
+        scopes: Optional[List[str]] = None,
+    ) -> Union['AzureCredentials', AnonymousCloudCredentials]:
+        if credentials is None:
+            if credentials_file:
+                return AzureCredentials.from_file(credentials_file, scopes=scopes)
+            return AzureCredentials.default_credentials(scopes=scopes)
+        if credentials_file is not None:
+            raise ValueError('Do not provide credentials_file and credentials.')
+        return credentials
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/python/hailtop/aiocloud/aioazure/credentials.py:74]: Done.

File: hail/python/hailtop/aiocloud/aioazure/fs.py (hunk 8 above)
Comment (danking) [reply hail/python/hailtop/aiocloud/aioazure/fs.py:373]: https://github.com/hail-is/hail/issues/14116

File: hail/python/hailtop/aiocloud/aiogoogle/client/bigquery_client.py (hunk 10 above)
Comment (danking) [reply hail/python/hailtop/aiocloud/aiogoogle/client/bigquery_client.py:114]: Mistake. Fixed.

**Comment by danking** [comment]
Several more lints were introduced into main in the meantime which needed to be fixed.

**Comment by danking** [comment]
I think this is ready for a fine look. It's been merged with the black change.

File: hail/python/hailtop/aiocloud/aioazure/credentials.py [hunk 17]
```diff
@@ -61,6 +61,15 @@ async def close(self) -> None:
 
 
 class AzureCredentials(CloudCredentials):
+    @staticmethod
+    def from_file_or_default(
+        credentials_file: Optional[str] = None,
+        scopes: Optional[List[str]] = None,
+    ) -> Union['AzureCredentials', AnonymousCloudCredentials]:
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiocloud/aioazure/credentials.py:68]: Where can this return anon credentials?

File: hail/python/hailtop/aiocloud/aiogoogle/client/bigquery_client.py [hunk 18]
```diff
@@ -104,8 +108,22 @@ async def __anext__(self):
...
+        credentials_file: Optional[str] = None,
+        params: Optional[Mapping[str, str]] = None,
+        **kwargs,
+    ):
+        super().__init__(
+            base_url=f'https://bigquery.googleapis.com/bigquery/v2/projects/{project}',
+            session=Session(
+                credentials=credentials or GoogleCredentials.from_file_or_default(credentials_file),
+                params=params,
+                **kwargs,
```
Comment (daniel-goldstein) [inline hail/python/hailtop/aiocloud/aiogoogle/client/bigquery_client.py:124]: Isn't this handled in the superclass?

File: hail/python/test/hail/fs/test_worker_driver_fs.py [hunk 19]
```diff
@@ -160,9 +161,13 @@ def test_can_access_public_blobs():
...
 @fails_local_backend
-def test_qob_can_use_sas_tokens():
+async def test_qob_can_use_sas_tokens():
     vcf = resource('sample.vcf')
-    account = AzureAsyncFS.parse_url(vcf).account
+    fs = AzureAsyncFS()
+    try:
+        account = fs.parse_url(vcf).account
+    finally:
+        await fs.close()
```
Comment (daniel-goldstein) [inline hail/python/test/hail/fs/test_worker_driver_fs.py:170]: I think this change is no longer needed since `parse_url` is now static

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
A few final bits

File: hail/python/hailtop/aiocloud/aioazure/credentials.py (hunk 17 above)
Comment (danking) [reply hail/python/hailtop/aiocloud/aioazure/credentials.py]: fixed

File: hail/python/hailtop/aiocloud/aiogoogle/client/bigquery_client.py (hunk 18 above)
Comment (danking) [reply hail/python/hailtop/aiocloud/aiogoogle/client/bigquery_client.py]: Removed.

File: hail/python/test/hail/fs/test_worker_driver_fs.py (hunk 19 above)
Comment (danking) [reply hail/python/test/hail/fs/test_worker_driver_fs.py]: done

**Review by daniel-goldstein — APPROVED** [review]
